src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -249,73 +271,66 @@ private void UpdateEndpoints()
-                        }
-                        if (allParameterPolicies == null)
-                        {
-                            allParameterPolicies = MvcEndpointInfo.BuildParameterPolicies(routePattern.Parameters, _parameterPolicyFactory);
-                        }
-
-                        // Replace parameter with literal value
-                        var parameterRouteValue = action.RouteValues[parameterPart.Name];
+                UpdatePathSegmentsWithLiterals(i, action, routePattern, newPathSegments, ref allParameterPolicies);
+            }
 
-                        // Route value could be null if it is a "known" route value.
-                        // Do not use the null value to de-normalize the route pattern,
-                        // instead leave the parameter unchanged.
-                        // e.g.
-                        //     RouteValues will contain a null "page" value if there are Razor pages
-                        //     Skip replacing the {page} parameter
-                        if (parameterRouteValue != null)
-                        {
-                            if (allParameterPolicies.TryGetValue(parameterPart.Name, out var parameterPolicies))
-                            {
-                                // Check if the parameter has a transformer policy
-                                // Use the first transformer policy
-                                for (var k = 0; k < parameterPolicies.Count; k++)
-                                {
-                                    if (parameterPolicies[k] is IOutboundParameterTransformer parameterTransformer)
-                                    {
-                                        parameterRouteValue = parameterTransformer.TransformOutbound(parameterRouteValue);
-                                        break;
-                                    }
-                                }
-                            }
+            IEnumerable<RoutePatternPathSegment> finalPathSegments = newPathSegments;
 
-                            segmentParts[j] = RoutePatternFactory.LiteralPart(parameterRouteValue);
-                        }
-                    }
+            if (requiredParameterEndpointRouteOrder != null)
+            {
+                // There are 3 cases where a parameter part can come into play:
+                // - /Home/{id?} - trailing single segment optional parameter
+                // - /Home.{id?} - trailing optional extension
+                // - /Home/{id?}/{catchAll*} - optional parameter then catch all
+
+                var hasCatchAll = newPathSegments.Last().IsSimple &&
+                    newPathSegments.Last().Parts.Last() is RoutePatternParameterPart parameterPart &&
+                    parameterPart.IsCatchAll;
+                var segmentIndex = hasCatchAll ? newPathSegments.Count - 2 : newPathSegments.Count - 1;
+                var segmentParts = newPathSegments[segmentIndex].Parts.ToList();
```

Comment:
ugh, some of these things make me gag a bit - worth making some private methods for this?

Reply:
Private methods? Do you mean to move it somewhere else or internal only methods on RoutePattern for mutating the pattern? Remember that route pattern is from routing and this is MVC land code.